src/mlpack/methods/reinforcement_learning/environment/bit_flipping.hpp

```diff
@@ -0,0 +1,258 @@
+   */
+  BitFlipping(const size_t maxSteps = 200,
+              const size_t length = 4) :
+      maxSteps(maxSteps),
+      length(length),
+      stepsPerformed(0)
+  { 
+    
+  }
+
+  /**
+   * Get reward and next state based on current
+   * state and current action.
+   *
+   * @param state The current state.
+   * @param action The current action.
+   * @param nextState The next state.
+   * @return reward, it's always 1.0.
+   */
+  double Sample(const State& state,
+                const Action& action,
+                State& nextState,
+                const State& transitionGoal)
+  {
+    // Update the number of steps performed.
+    stepsPerformed++;
+
+    // Modify state according to action
+    arma::colvec modifiedState = state.Data();
+    modifiedState(action.action) = 1 - modifiedState(action.action);
+    nextState.Data()= modifiedState;
+
+    // Check if the episode has terminated.
+    bool done = IsTerminal(nextState);
+
+    // Do not reward agent if it failed.
+    if (done)
+      return 1.0;
+
+    // Reward agent if it reaches the goal of transition and is not done
+    if (!done && sum(nextState.Data()) == sum(transitionGoal.Data()))
+    {
+      return 1.0;
+    }
+
+    return 0.0;
+  }
+
+  /**
+   * Get reward based on current state and current
+   * action.
+   *
+   * @param state The current state.
+   * @param action The current action.
+   * @return reward, it's always 1.0.
+   */
+  double Sample(const State& state, const Action& action, const State& transitionGoal)
+  {
+    State nextState;
+    return Sample(state, action, nextState, transitionGoal);
+  }
+
+  /**
+   * Initial state representation is randomly generated binary vector
+   *
+   * @return Initial state for each episode.
+   */
+  State InitialSample()
+  {
+    stepsPerformed = 0;
+    initialState = arma::randi<arma::colvec>(length, arma::distr_param(0, 1));
+    return State(initialState);
+  }
+
+  /**
+   * Get reward for particular goal
+   *
+   * @param nextState The next state.
+   * @param transitionGoal Transition's goal.
+   * @return Initial state for each episode.
+   */
+  double GetHERReward(const State& nextState,
+                    const State& transitionGoal)
+  {
+    // Reward agent if it reaches the goal of transition and is not done
+    if (sum(nextState.Data()) == sum(transitionGoal.Data()))
+    {
+      return 1.0;
+    }
+
+    return 0.0;
+  }
+
+  /**
+   * This function checks if the cart has reached the terminal state.
+   *
+   * @param state The desired state.
+   * @return true if state is a terminal state, otherwise false.
+   */
+  bool IsTerminal(const State& state) const
+  { 
+    arma::colvec currentState = state.Data();
+
+    if (maxSteps != 0 && stepsPerformed >= maxSteps)
+    {
+      Log::Info << "Episode terminated due to the maximum number of steps"
+          "being taken.";
+      return true;
+    }
+    else if (sum(currentState - goal) == 0)
```

Comment:
so, `currentState == goal`?

Reply:
It was giving some error i dont know why. Wait let me paste that here.
https://stackoverflow.com/questions/63062118/c-error-expression-must-have-bool-type

Reply:
Paste the error you get when `currentState == goal`

Reply:
> Paste the error you get when `currentState == goal`

I have pasted the stack overflow link for it. It was basically that
C++ Error: Expression Must Have Bool Type

Reply:
No, I meant compile it. Show me the entire compile error message. Or replicate this error in smallest possible code.

Reply:
Here is the pastebin link. - **https://pastebin.com/Nc7sFy1j**